lib/apiservers/engine/backends/volume.go

```diff
@@ -96,12 +105,103 @@ func (v *Volume) Volumes(filter string) ([]*types.Volume, []string, error) {
 		log.Infof("%s", vol.Name)
 		volumeMetadata, err := extractDockerMetadata(vol.Metadata)
 		if err != nil {
-			return nil, nil, fmt.Errorf("error unmarshalling docker metadata: %s", err)
+			return nil, nil, derr.NewErrorWithStatusCode(fmt.Errorf("error unmarshalling docker metadata: %s", err), http.StatusInternalServerError)
 		}
 		volume := NewVolumeModel(vol, volumeMetadata.Labels)
 		volumes = append(volumes, volume)
 	}
-	return volumes, nil, nil
+
+	// Parse and validate filter args
+	volumeFilters, err := filters.FromParam(filter)
+	if err != nil {
+		return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+	if err := volumeFilters.Validate(acceptedVolumeFilters); err != nil {
+		return nil, nil, derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	if volumeFilters.Len() == 0 {
+		return volumes, nil, nil
+	}
+
+	// If the dangling filter is specified, gather required items beforehand
+	dangling := false
+	joinedVolumes := make(map[string]struct{})
```

Comment:
Minor:  I think i understand what `joinedVolumes` are for, but it's possible (and likely) that I don't or will forget.  Can you please add a comment about what this is and what you're intending to do with it?